[dynamo] Simplify ODictGetItemSource index

anijain2305 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #117227

Goal is to keep the `index` of `ODictGetItemSource` capturable as Python value to help with Guard Tree refactor.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

anijain2305 committed:
[dynamo] Simplify ODictGetItemSource index

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo] Simplify ODictGetItemSource index"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

lezcano commented on torch/_dynamo/variables/user_defined.py:
```diff
@@ -845,11 +845,30 @@ def odict_getitem(self, tx, key):
+            raise ValueError("Trying to use a non-hashable key in a dict")
+
+        # TODO(anijain2305) Is there a better way to insert guard than iterating through hashable variable trackers?
+        if key.source:
```
What about tensors?

anijain2305 replied:
So, I don't know how to handle tensors here. The example I am dealing is this 

~~~
import torch
x = torch.zeros(3)
def nothing():
    pass

d = {
    1: torch.float32,
    nothing: torch.float32, # Force one of the keys to be not hashable
    x: torch.float32,
}

@torch.compile(backend="eager")
def f(x, y1):
    return torch.zeros(5, dtype=d[y1])

# f(torch.zeros(4), 1)
f(torch.zeros(4), x)
# f(torch.zeros(4), nothing)
~~~

Here `nothing` as key prevents the dict to be converted to `ConstDictVariable`. Instead it is represented as the `UserObjectVariable` tracker. Dynamo creates a `ODictGetItemSource` at `dtype=d[y1]` (odict is just a bad name, there is nothing about ordered dict here).  

Now if `y1` is actually a constant, no problem. 

But in the above example, I made it `TensorVariable`. The problem is what to put in the index of `ODictGetItem`. I want the index to be a PyObject so that I can just save that in the Guard Tree. I thought of using `ConstDictKeySource` by finding the order number of the key in the original dict. But I don't have the original tensor in the TensorVariable, just the `example_value`, so no way to find the order number in the original dict.

lezcano commented on torch/_dynamo/variables/user_defined.py:
```diff
@@ -845,11 +845,30 @@ def odict_getitem(self, tx, key):
 
         # TODO this should probably be merged with the dict handling
```
Wouldn't it be better to refactor OrderedDict into a subclass of `ConstantDictVariable`?

lezcano commented:
Also, welcome back :)

anijain2305 committed:
Update on "[dynamo] Simplify ODictGetItemSource index"


Goal is to keep the `index` of `ODictGetItemSource` capturable as Python value to help with Guard Tree refactor.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jansel commented:
Please re-request review once you have addressed @lezcano's comments and fixed tests

voznesenskym commented on torch/_dynamo/variables/user_defined.py:
```diff
@@ -840,16 +840,48 @@ def call_hasattr(self, tx, name: str) -> "VariableTracker":
     def odict_getitem(self, tx, key):
+        # This helper function is called for dicts or OrderedDicts, when atleast
+        # one of the keys is not hashable. If all the keys were hashable, the
+        # dict would have already been tracked as ConstDictVariable. Here, we
```
Hmm, if this is the invariant, this seems odd - perhaps its time to revisit ODictVariable vs ConstDictVariable vs DefaultDictVariable?

voznesenskym commented on torch/_dynamo/variables/user_defined.py:
```diff
@@ -840,16 +840,48 @@ def call_hasattr(self, tx, name: str) -> "VariableTracker":
+        if key.source and not is_hashable(key):
+            unimplemented("Non-hashable key in dict")
+
+        # Tensors are also considered hashable, but we don't support them here.
+        # There is no way to create a ConstDictKeySource here because we don't
+        # have the original tensor (example value is a fake tensor) to find the
+        # index in the dictionary.  This is probably ok because we care about
```
Have you seen `underlying_value` in SetVariable?

voznesenskym commented:
Discussed offline - the right way to get rid of the undesireable odict getitem method is to remove the weird odict_getitem and ODictGetItemSource (which are not what they say they are, anyway), improve the breadth of what we used to support in hashing, and collapse everything into ConstDictVariable.

Then, if we want a true ODictVariable for ordered dicts, we can do it via subclass.

lezcano commented:
The approach SGTM

lezcano commented:
but I'm not sure how will you be able to support a custom object in the dict key guards...
